index/scorch/merge.go

```diff
@@ -371,18 +374,25 @@ func (s *Scorch) planMergeAtSnapshot(ctx context.Context,
 		introStartTime := time.Now()
 		// it is safe to blockingly wait for the merge introduction
 		// here as the introducer is bound to handle the notify channel.
-		newSnapshot := <-sm.notify
+		introStatus := <-sm.notifyCh
 		introTime := uint64(time.Since(introStartTime))
 		atomic.AddUint64(&s.stats.TotFileMergeZapIntroductionTime, introTime)
 		if atomic.LoadUint64(&s.stats.MaxFileMergeZapIntroductionTime) < introTime {
 			atomic.StoreUint64(&s.stats.MaxFileMergeZapIntroductionTime, introTime)
 		}
 		atomic.AddUint64(&s.stats.TotFileMergeIntroductionsDone, 1)
-		if newSnapshot != nil {
-			_ = newSnapshot.DecRef()
+		if introStatus != nil && introStatus.indexSnapshot != nil {
+			_ = introStatus.indexSnapshot.DecRef()
+			if introStatus.skipped {
+				// decrement the ref counts on skipping introduction.
+				s.unmarkIneligibleForRemoval(filename)
+				_ = seg.DecRef()
```

Comment:
I found this like a bit misleading - so and had to look around the code a bit to understand why this is needed.
It makes sense - I feel seg.Close() would look better (on line 368 above too) although it's the same thing. I'll leave that up to you.

Reply:
done!

Reply:
Funny you should mention that.  

In the bluge codebase yesterday I just made a similar change.  We (IMHO) make the same mistake with *IndexSnapshot's all over the place as well.  We directly call DecRef(), when we should call Close() and let decRef() be an implementation detail of Close()

In the case of *IndexSnapshot's, I think the correct pattern is to have a single `getCurrent()` which uses RLock() to copy the root pointer, adds a ref, and returns it to you.  From that point, the caller simply uses and closes when done.  That way the ref counting is hidden from the user.

I'm not exactly sure that same pattern works for Segments, but I think the principals are the same.